src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -223,6 +223,44 @@ AcquireExecutorShardLock(Task *task, CmdType commandType)
 		LockShardResource(shardId, lockMode);
 	}
 
+	/*
+	 * If lock clause exists and it effects any reference table, we need to get
+	 * lock on shard resource. Type of lock is determined by the type of row lock
+	 * given in the query.
+	 */
+	if (task->rowLocksOnRelations != NIL)
+	{
+		ListCell *rowLockTypeCell = NULL;
+		ListCell *relationOidCell = NULL;
+		LOCKMODE rowLockMode = NoLock;
+
+		forboth(rowLockTypeCell, task->rowLocksOnRelations, relationOidCell,
+				task->tableOidList)
```

Comment:
In a similar situation (relationShardList), we opted for introducing a list of structs. Would it make sense to do the same thing here? It's a bit less brittle in terms of relying on the order of these lists being the same.